modules/core-components/src/test/java/org/mule/runtime/core/internal/source/polling/DefaultSchedulerMessageSourceTestCase.java

```diff
@@ -126,6 +129,80 @@ public String describeFailure() {
     });
   }
 
+  @Test
+  public void testStartWhenStopping() throws Exception {
+    DefaultSchedulerMessageSource schedulerMessageSource = createMessageSource();
+    SensingNullMessageProcessor flow = getSensingNullMessageProcessor();
+    schedulerMessageSource.setListener(flow);
+    schedulerMessageSource.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+
+    DefaultMuleContext spyMuleContext = spy((DefaultMuleContext) muleContext);
+    doReturn(true).when(spyMuleContext).isStopping();
+
+    schedulerMessageSource.setMuleContext(spyMuleContext);
+    schedulerMessageSource.start();
+
+    assertThat("Scheduler should not be started when MuleContext is stopping",
+               schedulerMessageSource.isStarted(), is(false));
+  }
+
+  @Test
+  public void testRunWhenStopping() throws Exception {
+    DefaultSchedulerMessageSource schedulerMessageSource = createMessageSource();
+    SensingNullMessageProcessor flow = getSensingNullMessageProcessor();
+    schedulerMessageSource.setListener(flow);
+    schedulerMessageSource.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+
+    DefaultMuleContext spyMuleContext = spy((DefaultMuleContext) muleContext);
+    doReturn(false).when(spyMuleContext).isStopping();
+    doReturn(true).when(spyMuleContext).isPrimaryPollingInstance();
+
+    schedulerMessageSource.setMuleContext(spyMuleContext);
+    schedulerMessageSource.start();
+    doReturn(true).when(spyMuleContext).isStopping();
+    new PollingProber(RECEIVE_TIMEOUT, 100).check(new Probe() {
+
+      @Override
+      public boolean isSatisfied() {
+        return flow.event == null;
+      }
+
+      @Override
+      public String describeFailure() {
+        return "flow event was processed when MuleContext was in stopping state";
+      }
+    });
+
+    verify(spyMuleContext, atLeastOnce()).isStopping();
+    verify(spyMuleContext, never()).isPrimaryPollingInstance();
+  }
+
+  @Test
+  public void testPollWhenStopping() throws Exception {
+    DefaultSchedulerMessageSource schedulerMessageSource = createMessageSource();
+    SensingNullMessageProcessor flow = getSensingNullMessageProcessor();
+    schedulerMessageSource.setListener(flow);
+    schedulerMessageSource.setAnnotations(singletonMap(LOCATION_KEY, TEST_CONNECTOR_LOCATION));
+
+    DefaultMuleContext spyMuleContext = spy((DefaultMuleContext) muleContext);
+    doReturn(true).when(spyMuleContext).isStopping();
+
+    schedulerMessageSource.setMuleContext(spyMuleContext);
+    schedulerMessageSource.trigger();
+    new PollingProber(RECEIVE_TIMEOUT, 100).check(new Probe() {
+
+      @Override
+      public boolean isSatisfied() {
+        return flow.event == null;
```

Comment:
Same as in the other test, what you need to assert is that the scheduled task finished, but it didn't trigger the flow (because of the "if" clause that you added)